packages/dds/tree/src/simple-tree/discrepancies.ts

```diff
@@ -321,87 +493,74 @@ function* trackObjectNodeDiscrepancies(
-				} satisfies FieldKindDiscrepancy;
-				break;
-			}
-			case "bExtra": {
-				if (result.value.kind === storedEmptyFieldSchema.kind) {
-					// In one of view/stored, this field is explicitly forbidden, but in the other it is implicitly forbidden
-					// (by way of omission). We treat these identically anyway.
-					break;
-				}
-				yield {
-					identifier,
-					fieldKey,
-					mismatch: "fieldKind",
-					view: storedEmptyFieldSchema.kind,
-					stored: result.value.kind,
-				} satisfies FieldKindDiscrepancy;
-				break;
-			}
-			case "both": {
-				yield* getFieldDiscrepancies(result.valueA, result.valueB, identifier, fieldKey);
-				break;
-			}
-			default:
-				break;
+	const viewKeys = new Set<FieldKey>();
+
+	for (const [_, { storedKey: fieldKey, schema: fieldSchema }] of view) {
+		const storedSchema = stored.objectNodeFields.get(fieldKey);
+		viewKeys.add(fieldKey);
+
+		// If the view schema has a field that's not in the stored schema
+		if (storedSchema === undefined) {
+			const viewKind =
+				convertFieldKind.get(fieldSchema.kind) ??
+				fail("A conversion from a FieldKind to a FlexFieldKind should exist");
+			yield {
+				identifier,
+				fieldKey,
+				mismatch: "fieldKind",
+				view: viewKind.identifier,
+				stored: storedEmptyFieldSchema.kind,
+			} satisfies FieldKindDiscrepancy;
+		} else {
+			yield* getFieldDiscrepancies(
+				fieldSchema,
+				storedSchema,
+				identifier,
+				fieldKey,
+				viewKindIsSequence,
+			);
 		}
 	}
-}
 
-function* compareMaps<K, V1, V2>(
-	a: ReadonlyMap<K, V1>,
-	b: ReadonlyMap<K, V2>,
-): Iterable<
-	| { type: "aExtra"; key: K; value: V1 }
-	| { type: "bExtra"; key: K; value: V2 }
-	| { type: "both"; key: K; valueA: V1; valueB: V2 }
-> {
-	for (const [key, valueA] of a) {
-		const valueB = b.get(key);
-		yield valueB === undefined
-			? { type: "aExtra", key, value: valueA }
-			: { type: "both", key, valueA, valueB };
-	}
-	for (const [key, valueB] of b) {
-		if (!a.has(key)) {
-			yield { type: "bExtra", key, value: valueB };
+	for (const [fieldKey, schema] of stored.objectNodeFields) {
+		// If the stored schema has a field that's not in the view schema
+		if (!viewKeys.has(fieldKey)) {
+			if (schema.kind === forbiddenFieldKindIdentifier) {
+				// In the stored schema the field is explicitly forbidden.
+				// This has the same semantics of the field not being mentioned in the stored schema,
+				// and thus is compatible with the view schema which does not mention this field.
+				continue;
+			}
+			yield {
+				identifier,
+				fieldKey,
+				mismatch: "fieldKind",
+				view: storedEmptyFieldSchema.kind,
+				stored: schema.kind,
+			} satisfies FieldKindDiscrepancy;
 		}
 	}
 }
+
 /**
- * @remarks
- *
  * This function uses incompatibilities to determine if changes to a document schema are backward-compatible, i.e., it determines
  * whether the `view` schema allows a superset of the documents that the `stored` schema allows.
```

Comment:
We should use consistent terminology (like discrepancies instead of incompatibilities, stored schema instead of document schema, canUpgrade instead of backward-compatible).


```suggestion
 * This function uses discrepancies to determine if replacing the provided stored schema to a stored schema derived from the provided view schema would support a superset of the documents permitted by the provided stored schema.
```

Note that this phrasing suggests to me that a better way to implement this would be to use toStoredSchema can the stored schema isRepoSuperset.


Given that this function only exists to test the other utilities here, I think such an implementation, and moving it into the test code, would make more sense than redundantly implementing canUpgrade using roughly the same approach as the production.